File: examples/tests/vp/test_mpp_encoder.cpp
#include "module/vi/module_memReader.hpp"
#include "module/vp/module_mppenc.hpp"
#include "module/vp/module_rga.hpp"
#include "tests/module_test_utils.hpp"

#include <algorithm>
#include <atomic>
#include <cstdio>
#include <memory>

using namespace FFMedia;
using namespace FFMediaTest;

namespace
{
void usage(const char* program)
{
    ff_info("Usage: %s OUTPUT [options]\n"
            "  --codec h264|h265|mjpeg --width 1920 --height 1080\n"
            "  --format nv12 [--hstride N --vstride N]\n"
            "  --fps 30 --gop 60 --bps 2048 --rc cbr|vbr|fixqp|avbr\n"
            "  --quality 0.8 --profile baseline|main|high\n"
            "  --duration-us N --output-timeout-ms -1 --cache-frames 0\n"
            "  --intra-refresh --refresh-mode 0 --refresh-num 10\n"
            "  --parallel-buffers 0   Insert RGA queue for parallel encoding\n"
            "  --buffer-type noncache|cache|malloc|dma32|cache-dma32\n"
            "  --frames 300 --wait-timeout-ms 2000 --drain-timeout-ms 2000 --animate\n"
            "  --report-every 100 --verbose --external-consumer --dump-pipe\n",
            program);
}

bool parseCodec(const std::string& name, media_codec_t& codec)
{
    if (name == "h264")
        codec = MEDIA_CODEC_VIDEO_H264;
    else if (name == "h265")
        codec = MEDIA_CODEC_VIDEO_H265;
    else if (name == "mjpeg")
        codec = MEDIA_CODEC_VIDEO_MJPEG;
    else
        return false;
    return true;
}

bool parseRcMode(const std::string& name, EncodeRcMode& mode)
{
    if (name == "cbr")
        mode = ENCODE_RC_MODE_CBR;
    else if (name == "vbr")
        mode = ENCODE_RC_MODE_VBR;
    else if (name == "fixqp")
        mode = ENCODE_RC_MODE_FIXQP;
    else if (name == "avbr")
        mode = ENCODE_RC_MODE_AVBR;
    else
        return false;
    return true;
}

bool parseProfile(const std::string& name, EncodeProfile& profile)
{
    if (name == "baseline")
        profile = ENCODE_PROFILE_BASELINE;
    else if (name == "main")
        profile = ENCODE_PROFILE_MAIN;
    else if (name == "high")
        profile = ENCODE_PROFILE_HIGH;
    else
        return false;
    return true;
}

bool parseBufferType(const std::string& name, VideoBuffer::BUFFER_TYPE& type)
{
    if (name == "noncache")
        type = VideoBuffer::DRM_BUFFER_NONCACHEABLE;
    else if (name == "cache")
        type = VideoBuffer::DRM_BUFFER_CACHEABLE;
    else if (name == "malloc")
        type = VideoBuffer::MALLOC_BUFFER;
    else if (name == "dma32")
        type = VideoBuffer::DRM_BUFFER_NONCACHEABLE_DMA32;
    else if (name == "cache-dma32")
        type = VideoBuffer::DRM_BUFFER_CACHEABLE_DMA32;
    else
        return false;
    return true;
}
}  // namespace

int main(int argc, char** argv)
{
    CliOptions options(argc, argv);
    if (options.has("help") || options.positional().empty()) {
        usage(argv[0]);
        return options.has("help") ? 0 : 2;
    }

    media_codec_t codec = MEDIA_CODEC_VIDEO_H265;
    EncodeRcMode rc_mode = ENCODE_RC_MODE_CBR;
    EncodeProfile profile = ENCODE_PROFILE_HIGH;
    VideoBuffer::BUFFER_TYPE buffer_type = VideoBuffer::DRM_BUFFER_NONCACHEABLE;
    if (!parseCodec(options.get("codec", "h265"), codec) || !parseRcMode(options.get("rc", "cbr"), rc_mode) || !parseProfile(options.get("profile", "high"), profile) || !parseBufferType(options.get("buffer-type", "noncache"), buffer_type)) {
        ff_error("Invalid codec, rate-control mode, profile, or buffer type\n");
        return 2;
    }

    const uint32_t width = options.getLong("width", 1920);
    const uint32_t height = options.getLong("height", 1080);
    const uint32_t format = v4l2GetFmtByName(options.get("format", "nv12").c_str());
    if (!width || !height || !format) {
        ff_error("Invalid input dimensions or format\n");
        return 2;
    }
    const ImagePara input_para(width, height,
                               options.getLong("hstride", width),
                               options.getLong("vstride", height), format);

    FILE* output_file = std::fopen(options.positional()[0].c_str(), "wb");
    if (!output_file) {
        ff_error("Failed to open output file: %s\n", options.positional()[0].c_str());
        return 1;
    }

    auto input_buffer = std::make_shared<VideoBuffer>(buffer_type);
    input_buffer->allocBuffer(input_para);
    if (!input_buffer->getActiveData()) {
        ff_error("Failed to allocate input buffer\n");
        std::fclose(output_file);
        return 1;
    }
    input_buffer->fillWithColor(32, 128, 224);
    if (buffer_type == VideoBuffer::DRM_BUFFER_CACHEABLE || buffer_type == VideoBuffer::DRM_BUFFER_CACHEABLE_DMA32)
        input_buffer->flushDrmBuf();

    auto reader = std::make_shared<ModuleMemReader>(input_para);
    int ret = reader->init();
    if (ret < 0) {
        ff_error("Failed to init memory reader: %d\n", ret);
        std::fclose(output_file);
        return ret;
    }
    dumpOutputMediaChannels(*reader);

    std::shared_ptr<ModuleMedia> producer = reader;
    std::shared_ptr<ModuleRga> rga;
    const int parallel_buffers = options.getLong("parallel-buffers", 0);
    const int cache_frames = options.getLong("cache-frames", 0);
    if (cache_frames > 1 && parallel_buffers <= cache_frames) {
        ff_error("cache-frames must be smaller than parallel-buffers when greater than 1\n");
        std::fclose(output_file);
        return 2;
    }
    if (parallel_buffers > 0) {
        rga = std::make_shared<ModuleRga>(input_para, RGA_ROTATE_NONE);
        ret = rga->connectProducer(reader);
        if (ret < 0) {
            ff_error("Failed to connect RGA: %d\n", ret);
            std::fclose(output_file);
            return ret;
        }
        rga->setBufferCount(parallel_buffers);
        ret = rga->init();
        if (ret < 0) {
            ff_error("Failed to init RGA: %d\n", ret);
            std::fclose(output_file);
            return ret;
        }
        dumpOutputMediaChannels(*rga);
        producer = rga;
    }

    const int fps = options.getLong("fps", 30);
    const int gop = options.getLong("gop", 60);
    const int bps = options.getLong("bps", 2048);
    const float quality = options.getDouble("quality", 0.8);
    auto encoder = std::make_shared<ModuleMppEnc>(codec, fps, gop, bps,
                                                  rc_mode, quality, profile);
    ret = encoder->connectProducer(producer);
    if (ret < 0) {
        ff_error("Failed to connect encoder: %d\n", ret);
        std::fclose(output_file);
        return ret;
    }
    encoder->setBufferCount(options.getLong("buffers", 8));
    encoder->setOutputTimeOut(options.getLong("output-timeout-ms", -1));
    encoder->setInputCachePoolSize(cache_frames);
    if (options.has("duration-us"))
        encoder->setDuration(options.getLong("duration-us", 0));
    if (options.has("intra-refresh")) {
        encoder->setIntraRefresh(true, options.getLong("refresh-mode", 0),
                                 options.getLong("refresh-num", 10));
    }

    const uint64_t max_frames = options.getLong("frames", 300);
    RunMonitor monitor(max_frames, options.getDouble("duration", 0.0),
                       options.getLong("report-every", 100),
                       options.getBool("verbose"));
    installSignalHandlers(monitor);
    std::atomic_bool write_failed(false);
    reader->setMediaStatusChangeHooker(
        [&monitor](const std::string& name, MediaStatus status) {
            monitor.onStatus(name, status);
        });
    encoder->setMediaStatusChangeHooker(
        [&monitor](const std::string& name, MediaStatus status) {
            monitor.onStatus(name, status);
        });
    encoder->setMediaBufferProduceHooker(
        [&monitor, output_file, &write_failed](const std::string& name,
                                               int queue_size,
                                               std::shared_ptr<MediaBuffer> buffer) {
            if (buffer && std::fwrite(buffer->getActiveData(), 1, buffer->getActiveSize(), output_file) != buffer->getActiveSize()) {
                write_failed = true;
                monitor.requestStop();
            }
            monitor.onBuffer(name, queue_size, buffer);
        });

    ret = encoder->init();
    if (ret < 0) {
        ff_error("Failed to init MPP encoder: %d\n", ret);
        std::fclose(output_file);
        return ret;
    }
    dumpOutputMediaChannels(*encoder);
    dumpExtraBuffer("encoder", encoder->getExtraBuffer());

    std::shared_ptr<ModuleMedia> external_consumer;
    std::atomic<uint64_t> external_frames(0);
    if (options.has("external-consumer")) {
        external_consumer = encoder->addExternalConsumer(
            "encoder-external", [&external_frames](const std::string&, int,
                                                   std::shared_ptr<MediaBuffer>) {
                ++external_frames;
            });
    }
    if (options.has("dump-pipe"))
        reader->dumpPipe();

    monitor.reset();
    reader->start();
    const int wait_timeout = options.getLong("wait-timeout-ms", 2000);
    uint64_t input_frames = 0;
    while (!monitor.stopped() && (!max_frames || input_frames < max_frames)) {
        if (options.getBool("animate")) {
            input_buffer->fillWithColor(input_frames & 0xff,
                                        (input_frames * 3) & 0xff,
                                        (input_frames * 7) & 0xff);
            if (buffer_type == VideoBuffer::DRM_BUFFER_CACHEABLE || buffer_type == VideoBuffer::DRM_BUFFER_CACHEABLE_DMA32)
                input_buffer->flushDrmBuf();
        }
        input_buffer->setPUstimestamp(fps > 0 ? input_frames * 1000000 / fps : 0);
        ret = reader->setInputBuffer(input_buffer);
        if (ret != 0) {
            ff_error("setInputBuffer failed: %d\n", ret);
            break;
        }
        ret = reader->waitProcess(wait_timeout);
        if (ret != 0) {
            ff_error("waitProcess failed or timed out: %d\n", ret);
            break;
        }
        ++input_frames;
    }

    const auto drain_deadline = std::chrono::steady_clock::now() + std::chrono::milliseconds(options.getLong("drain-timeout-ms", 2000));
    while (ret == 0 && monitor.frames() < input_frames && std::chrono::steady_clock::now() < drain_deadline && !monitor.abnormal()) {
        std::this_thread::sleep_for(std::chrono::milliseconds(10));
    }
    const bool incomplete_output = monitor.frames() != input_frames;
    if (incomplete_output)
        ff_warn("Encoder drain incomplete: input %lu, output %lu\n",
                input_frames, monitor.frames());

    reader->setProcessStatus(ModuleMemReader::PROCESS_STATUS_EXIT);
    encoder->stop();
    if (rga)
        rga->stop();
    reader->stop();
    std::fclose(output_file);
    if (options.has("dump-pipe"))
        reader->dumpPipeSummary();
    monitor.printSummary("MPP encoder");
    ff_info("Input frames=%lu, output frames=%lu\n", input_frames,
            monitor.frames());
    if (external_consumer)
        ff_info("External consumer received %lu buffers\n", external_frames.load());
    return (ret != 0 || write_failed || monitor.abnormal() || incomplete_output) ? 1 : 0;
}

File: examples/tests/vp/test_mpp_encoder.py
#!/usr/bin/env python3

import argparse
import os
import sys
import time

sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

import ff_pymedia as ff
from module_test_utils import RunMonitor, dump_output_channels


CODECS = {
    "h264": ff.MEDIA_CODEC_VIDEO_H264,
    "h265": ff.MEDIA_CODEC_VIDEO_H265,
    "mjpeg": ff.MEDIA_CODEC_VIDEO_MJPEG,
}
RC_MODES = {
    "cbr": ff.ENCODE_RC_MODE_CBR,
    "vbr": ff.ENCODE_RC_MODE_VBR,
    "fixqp": ff.ENCODE_RC_MODE_FIXQP,
    "avbr": ff.ENCODE_RC_MODE_AVBR,
}
PROFILES = {
    "baseline": ff.ENCODE_PROFILE_BASELINE,
    "main": ff.ENCODE_PROFILE_MAIN,
    "high": ff.ENCODE_PROFILE_HIGH,
}
BUFFER_TYPES = {
    "noncache": ff.DRM_BUFFER_NONCACHEABLE,
    "cache": ff.DRM_BUFFER_CACHEABLE,
    "malloc": ff.MALLOC_BUFFER,
    "dma32": ff.DRM_BUFFER_NONCACHEABLE_DMA32,
    "cache-dma32": ff.DRM_BUFFER_CACHEABLE_DMA32,
}


def parse_args():
    parser = argparse.ArgumentParser(description="Full ModuleMppEnc usage test")
    parser.add_argument("output")
    parser.add_argument("--codec", choices=CODECS, default="h265")
    parser.add_argument("--width", type=int, default=1920)
    parser.add_argument("--height", type=int, default=1080)
    parser.add_argument("--hstride", type=int)
    parser.add_argument("--vstride", type=int)
    parser.add_argument("--format", default="nv12")
    parser.add_argument("--buffer-type", choices=BUFFER_TYPES, default="noncache")
    parser.add_argument("--fps", type=int, default=30)
    parser.add_argument("--gop", type=int, default=60)
    parser.add_argument("--bps", type=int, default=2048)
    parser.add_argument("--rc", choices=RC_MODES, default="cbr")
    parser.add_argument("--quality", type=float, default=0.8)
    parser.add_argument("--profile", choices=PROFILES, default="high")
    parser.add_argument("--duration-us", type=int)
    parser.add_argument("--output-timeout-ms", type=int, default=-1)
    parser.add_argument("--cache-frames", type=int, default=0)
    parser.add_argument("--intra-refresh", action="store_true")
    parser.add_argument("--refresh-mode", type=int, choices=(0, 1), default=0)
    parser.add_argument("--refresh-num", type=int, default=10)
    parser.add_argument("--parallel-buffers", type=int, default=0)
    parser.add_argument("--buffers", type=int, default=8)
    parser.add_argument("--frames", type=int, default=300)
    parser.add_argument("--duration", type=float, default=0.0)
    parser.add_argument("--wait-timeout-ms", type=int, default=2000)
    parser.add_argument("--drain-timeout-ms", type=int, default=2000)
    parser.add_argument("--animate", action="store_true")
    parser.add_argument("--report-every", type=int, default=100)
    parser.add_argument("--verbose", action="store_true")
    parser.add_argument("--external-consumer", action="store_true")
    parser.add_argument("--dump-pipe", action="store_true")
    return parser.parse_args()


def main():
    args = parse_args()
    if args.cache_frames > 1 and args.parallel_buffers <= args.cache_frames:
        print("cache-frames must be smaller than parallel-buffers when greater than 1")
        return 2
    pixel_format = ff.v4l2GetFmtByName(args.format)
    if not pixel_format or args.width <= 0 or args.height <= 0:
        print("Invalid dimensions or pixel format")
        return 2
    image_para = ff.ImagePara(
        args.width, args.height, args.hstride or args.width,
        args.vstride or args.height, pixel_format
    )
    input_buffer = ff.VideoBuffer(BUFFER_TYPES[args.buffer_type])
    input_buffer.allocBuffer(image_para)
    if input_buffer.getSize() == 0:
        print("Failed to allocate input buffer")
        return 1
    input_buffer.fillWithColor(32, 128, 224)
    if args.buffer_type in ("cache", "cache-dma32"):
        input_buffer.flushDrmBuf()

    reader = ff.ModuleMemReader(image_para)
    ret = reader.init()
    if ret < 0:
        print(f"Failed to init memory reader: {ret}")
        return 1
    dump_output_channels(reader)

    producer = reader
    rga = None
    if args.parallel_buffers > 0:
        rga = ff.ModuleRga(image_para, ff.RGA_ROTATE_NONE)
        ret = rga.connectProducer(reader)
        if ret < 0:
            print(f"Failed to connect RGA: {ret}")
            return 1
        rga.setBufferCount(args.parallel_buffers)
        ret = rga.init()
        if ret < 0:
            print(f"Failed to init RGA: {ret}")
            return 1
        dump_output_channels(rga)
        producer = rga

    encoder = ff.ModuleMppEnc(
        CODECS[args.codec], args.fps, args.gop, args.bps,
        RC_MODES[args.rc], args.quality, PROFILES[args.profile]
    )
    ret = encoder.connectProducer(producer)
    if ret < 0:
        print(f"Failed to connect encoder: {ret}")
        return 1
    encoder.setBufferCount(args.buffers)
    encoder.setOutputTimeOut(args.output_timeout_ms)
    encoder.setInputCachePoolSize(args.cache_frames)
    if args.duration_us is not None:
        encoder.setDuration(args.duration_us)
    if args.intra_refresh:
        encoder.setIntraRefresh(True, args.refresh_mode, args.refresh_num)

    monitor = RunMonitor(args.frames, args.duration, args.report_every, args.verbose)
    monitor.install_signal_handlers()
    reader.setMediaStatusChangeHooker(monitor.status_callback)
    encoder.setMediaStatusChangeHooker(monitor.status_callback)
    output_file = open(args.output, "wb")

    def output_callback(name, queue_size, buffer):
        output_file.write(buffer.getActiveData())
        monitor.output_callback(name, queue_size, buffer)

    encoder.setMediaBufferProduceHooker(output_callback)
    ret = encoder.init()
    if ret < 0:
        output_file.close()
        print(f"Failed to init encoder: {ret}")
        return 1
    dump_output_channels(encoder)
    extra = encoder.getExtraBuffer()
    print(f"Encoder extra data: {extra.getActiveSize() if extra else 0} bytes")

    external = None
    external_frames = [0]
    if args.external_consumer:
        def external_callback(_name, _queue_size, _buffer):
            external_frames[0] += 1
        external = encoder.addExternalConsumer("encoder-external", external_callback)
    if args.dump_pipe:
        reader.dumpPipe()

    monitor.reset()
    reader.start()
    input_frames = 0
    start_time = time.monotonic()
    while not monitor.stop_event.is_set() and (args.frames == 0 or input_frames < args.frames):
        if args.duration > 0 and time.monotonic() - start_time >= args.duration:
            break
        if args.animate:
            input_buffer.fillWithColor(
                input_frames & 0xFF, (input_frames * 3) & 0xFF,
                (input_frames * 7) & 0xFF
            )
            if args.buffer_type in ("cache", "cache-dma32"):
                input_buffer.flushDrmBuf()
        input_buffer.setPUstimestamp(input_frames * 1_000_000 // args.fps if args.fps else 0)
        ret = reader.setInputBuffer(input_buffer)
        if ret != 0:
            print(f"setInputBuffer failed: {ret}")
            break
        ret = reader.waitProcess(args.wait_timeout_ms)
        if ret != 0:
            print(f"waitProcess failed or timed out: {ret}")
            break
        input_frames += 1

    drain_deadline = time.monotonic() + args.drain_timeout_ms / 1000.0
    while ret == 0 and monitor.frames < input_frames and time.monotonic() < drain_deadline:
        time.sleep(0.01)
    incomplete_output = monitor.frames != input_frames
    if incomplete_output:
        print(f"Encoder drain incomplete: input {input_frames}, output {monitor.frames}")

    reader.setProcessStatus(ff.PROCESS_STATUS_EXIT)
    encoder.stop()
    if rga is not None:
        rga.stop()
    reader.stop()
    output_file.close()
    if args.dump_pipe:
        reader.dumpPipeSummary()
    monitor.print_summary("MPP encoder")
    print(f"Input frames={input_frames}, output frames={monitor.frames}")
    if external is not None:
        print(f"External consumer received {external_frames[0]} buffers")
    return 1 if ret != 0 or monitor.abnormal or incomplete_output else 0


if __name__ == "__main__":
    raise SystemExit(main())
